src/transport/SessionManager.cpp

```diff
@@ -743,32 +728,6 @@ CHIP_ERROR SessionManager::TCPConnect(const PeerAddress & peerAddress, Transport
 
     return CHIP_ERROR_INCORRECT_STATE;
 }
-
-CHIP_ERROR SessionManager::TCPDisconnect(const PeerAddress & peerAddress)
-{
-    if (mTransportMgr != nullptr)
-    {
-        char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
-        peerAddress.ToString(peerAddrBuf);
-        ChipLogProgress(Inet, "Disconnecting TCP connection from peer at %s.", peerAddrBuf);
-        mTransportMgr->TCPDisconnect(peerAddress);
-    }
-
-    return CHIP_NO_ERROR;
-}
-
-void SessionManager::TCPDisconnect(Transport::ActiveTCPConnectionState * conn, bool shouldAbort)
-{
-    if (mTransportMgr != nullptr && conn != nullptr)
-    {
-        char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
-        conn->mPeerAddr.ToString(peerAddrBuf);
-        ChipLogProgress(Inet, "Disconnecting TCP connection from peer at %s.", peerAddrBuf);
-        mTransportMgr->TCPDisconnect(conn, shouldAbort);
-
-        MarkSecureSessionOverTCPForEviction(conn, CHIP_NO_ERROR);
```

Comment:
Earlier, explicitly calling this TCPDisconnect(conn, ..) also marked the session for eviction. By using the automatic release and invocation of the TCPDisconnect at the transport level is, potentially, missing the marking of the corresponding session for eviction, leaving an active session without any underlying connection.
We would need a way for the TCPDisconnect(..) to signal the SessionManager to mark the session as such.

Reply:
We really should never have been explicitly closing the TCP connection - as we share the connection, no single user should be able to force it closed

If the session is closed organically, say the network dies, then we will continue to have the close callback which automatically marks all sessions for eviction (`HandleConnectionClosed`)

The one exception to this is testing - the easiest way to test that sessions were properly evicted is by forcing the connection closed, which is the reason `TCP_Tests.py` was failing; I added a `ForceDisconnect` to allow us to test for this case, but it really should never be used in production